node/src/main/kotlin/net/corda/node/services/persistence/NodeAttachmentService.kt

```
@@ -324,6 +326,21 @@ class NodeAttachmentService(
                     val jarSigners = getSigners(bytes)
                     val contractVersion = getVersion(bytes)
                     val session = currentDBSession()
+
+                    contractClassNames.forEach {
+                        val existingContractsImplementations = queryAttachments(AttachmentQueryCriteria.AttachmentsQueryCriteria(
+                                contractClassNamesCondition = Builder.equal(listOf(it)),
+                                versionCondition = Builder.equal(contractVersion),
+                                isSignedCondition = Builder.equal(jarSigners.isNotEmpty()))
+                        )
+                        if (existingContractsImplementations.isNotEmpty()) {
+                            if (jarSigners.isNotEmpty())
+                                throw DuplicateContractClassException(it, contractVersion, existingContractsImplementations.map { it.toString() })
+                            else
+                                log.warn("Importing contract $it version '$contractVersion' from unsigned attachment which is already present in unsigned attachment(s) $existingContractsImplementations")
```

I think we should also fail fast on duplicate signed jars where `targetPlatformVersion >= 4`.